tools/unirestore/README.md

```diff
@@ -0,0 +1,99 @@
+# Universal Restore Utility
+
+Restore a backup of a set of SSTable files taken from any given Scylla/Cassandra cluster to any size/topology cluster. 
+
+
+Overview of what it does and how it does it:
+
+As a user, I have a set of sstable files I have backed up from my old cluster. The set looks like this[1]:
+
+`root-dir/host/keyspace/table/sstable_files`
+
+Also, the user provides me with:
+
+- the original schema dump in a CQL file (can be sanitized to include only keyspaces/tables)
+- ssh credentials for a sudo user on all nodes (key and password access are supported)
+- One of the node IPs in the cluster for cluster metadata gathering
+- If needed, API URL and CQLSH credentials
+
+I have a new cluster into which I'd like to load the data. The node count and topology is irrelevant (though if the new cluster 
+is multi-DC based, it would make sense to limit the destination node list to a local DC and adjust the replicas after restore, 
+in order to avoid cross DC traffic)
+
+I assume the schema has already been loaded into the new cluster.
+
+## The workflow:
+
+1. Parse the original schema dump, to have a list of valid keyspaces and tables.
+2. Extract the cluster node IP list. 
+3. Ensure schema versions are up to date, and in sync across the cluster.
+4. Extract the token ring data for each valid keyspace [temporarily unavailable]
+
+5. Find the correct destination paths for each ks/table (use system tables to avoid looking at dropped tables)
+  5.1 Ensure the destination dirs are writeable
+6. Build restore plan
+  6.1 organize source files as bundles of 9 sstable files with source paths
+  6.2 [temporarily unavailable] add the first/last tokens set for each bundle
+  6.3 [temporarily unavailable] add the list of hosts for each keyspace where the file token range overlaps the node/ks token range
+  6.4 add destination path (ks/table/upload) to each bundle
+7. Iterate over the bundled file sets, go one source host per iteration, copying the largest bundles first.
+  7.1 Measure the bundle (9 sstable files) size
+  7.2 Check `df` on the destination nodes
+    7.2.1 Confirm There's room for the incoming bundle
+    7.2.2 Make sure the preset percentage of free space is not exceeded
+    7.2.3 If disk space is low or below the threshold
+      7.2.3.1 Send the API call to stop compactions
+      7.2.3.2 Set the correct file permissions on the data in the upload dirs
+      7.2.3.3 Call nodetool refresh[2], and wait for it to finish[3]
+  7.3 If enough free space is available copy the 9-file bundle into all the relevant nodes in parallel to save time[4]
+  7.4 Mark bundle as complete in the restore plan file, so that if something fails, the script can be resumed.
+  7.5 When all the files from a source host have been restored, call a refresh.[5]
```

Comment:
Missing: need to set gc_grace_seconds to a very large value while this is going on. If you don't, then you might load an sstable with an old tombstone, get it compact away, then load an sstable with a younger cell that is covered by that tombstone. The result is data resurrection.

After the process is complete, need to set gc_grace_seconds back.

Reply:
What about calling `nodetool cleanup`?

Reply:
Cleanup is being called as well. I didn't mention it explicitly (I probably should)

About gc_grace)seconds change - is this something that can be done on the fly (via the API maybe?)
Because if not, I think it would be a better idea to simply document this requirement and tell the user to set it high, do a rolling restart, and then run the restore tool. I don't think this tool should be managing rolling restarts on it's own.

Reply:
One more thing wrt the resurrection issue: is this an issue? My point is, I am restoring from a backup, this procedure is supposed to restore lost/deleted data. By setting gc_grace_seconds very high, I might be shooting the user in the foot here. 

In short, I think it's best to simply document this snag and let the user decide whether they want to play with gc_grace_seconds before a restore or not. @avikivity @slivne @vladzcloudius

Reply:
gc_grace_seconds can be changed with ALTER TABLE.

Reply:
data resurrection is equivalent to a lost delete. That is, the sequence

    INSERT INTO tab (key) VALUES (1);
    DELETE FROM tab WHERE key = 1;

can result in a partition with key=1, as if the delete was never issued. It is quite likely to happen during restore from an old backup, as most tombstones will be old.

I don't think it can be documented as a snag. How would you describe it? Sometimes, rows that were deleted in the backup appear in the restore?

Reply:
If I deleted a row and my backup contains it, the reason for the restore might very well be to bring back the data I accidentally deleted. 

as for gc_grace changes, I already added it to the TODOs for the next patch